server/embedded/src/org/labkey/embedded/LabKeyTomcatServletWebServerFactory.java

```diff
@@ -0,0 +1,521 @@
+                        {
+                            extractZip(distInputStream, destDirectory);
+                        }
+                    }
+                }
+
+                if (!foundDistributionZip)
+                {
+                    throw new ConfigException("Unable to find distribution zip required to run LabKey server.");
+                }
+            }
+        }
+        catch (IOException | ConfigException e)
+        {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static String getExecutableJar(String currentPath) throws ConfigException
+    {
+        File currentDir = new File(currentPath);
+        List<String> jarsPresent = new ArrayList<>();
+
+        File[] files = currentDir.listFiles();
+        if (files != null)
+        {
+            for (File file : files)
+            {
+                if (file.getName().toLowerCase().endsWith(".jar"))
+                {
+                    jarsPresent.add(file.getName());
+                }
+            }
+        }
+
+        if (jarsPresent.isEmpty())
+        {
+            throw new ConfigException("Executable jar not found.");
+        }
+
+        // only 1 jar should be there
+        if (jarsPresent.size() == 1)
+        {
+            return jarsPresent.get(0);
+        }
+
+        throw new ConfigException("Multiple jars found - " + jarsPresent + ". Must provide only one jar.");
+    }
+
+    private void extractZip(InputStream zipInputStream, String destDirectory) throws IOException
+    {
+        File destDir = new File(destDirectory);
+        //noinspection SSBasedInspection
+        if (!destDir.exists() && !destDir.mkdirs())
+        {
+            throw new IOException("Failed to create directory " + destDir + " - please check file system permissions");
+        }
+        try (ZipInputStream zipIn = new ZipInputStream(zipInputStream))
+        {
+            ZipEntry entry = zipIn.getNextEntry();
+            // iterates over entries in the zip file
+            while (entry != null)
+            {
+                String filePath = destDirectory + File.separator + entry.getName();
+                if (!entry.isDirectory())
+                {
+                    // if the entry is a file, extracts it
+                    extractFile(zipIn, filePath);
+                }
+                else
+                {
+                    // if the entry is a directory, make the directory
+                    File dir = new File(filePath);
+                    //noinspection SSBasedInspection
+                    if (!dir.exists() && !dir.mkdirs())
+                    {
+                        throw new IOException("Failed to create directory " + dir + " - please check file system permissions");
+                    }
+                }
+                zipIn.closeEntry();
+                entry = zipIn.getNextEntry();
+            }
+        }
+    }
+
+    private static final int BUFFER_SIZE = 4096;
```

Comment:
Should probably move this up with other constants

Reply:
I moved it. I sometimes find it helpful to keep the declaration close to the usage, but also agree that sprinkling them throughout the class isn't a good option